src/cdomain/value/cdomains/int/bitfieldDomain.ml

```diff
@@ -0,0 +1,623 @@
+      let shift_counts = concretize (z2, o2) in
+      List.fold_left (fun acc c ->
+          let next = shift_right ik bf c in
+          join acc next
+        ) (zero_mask, zero_mask) shift_counts
+
+  let shift_left _ (z,o) c =
+    let zero_mask = bitmask_up_to c in
+    ((z <<: c) |: zero_mask, o <<: c)
+
+  let shift_left ik bf (z2, o2) = 
+    if is_const (z2, o2) 
+    then 
+      shift_left ik bf (Ints_t.to_int o2)
+    else
+      let shift_counts = concretize (z2, o2) in
+      List.fold_left (fun acc c ->
+          let next = shift_left ik bf c in
+          join acc next
+        ) (zero_mask, zero_mask) shift_counts
+
+  let nth_bit p n = if nth_bit p n then Ints_t.one else Ints_t.zero
+end
+
+module BitfieldFunctor (Ints_t : IntOps.IntOps): SOverflow with type int_t = Ints_t.t and type t = (Ints_t.t * Ints_t.t) = struct
+
+  include InfixIntOps (Ints_t)
+
+  let name () = "bitfield"
+  type int_t = Ints_t.t
+  type t = (Ints_t.t * Ints_t.t) [@@deriving eq, ord, hash]
+
+  module BArith = BitfieldArith (Ints_t)
+
+  let top () = (BArith.one_mask, BArith.one_mask)
+  let bot () = (BArith.zero_mask, BArith.zero_mask)
+
+  let top_of ik = 
+    if isSigned ik then top () 
+    else (BArith.one_mask, Ints_t.of_bigint (snd (Size.range ik)))
+
+  let bot_of ik = bot ()
+
+  let to_pretty_bits (z,o) = 
+    let known_bitmask = ref (BArith.bits_known (z,o)) in
+    let invalid_bitmask = ref (BArith.bits_invalid (z,o)) in
+    let o_mask = ref o in
+    let z_mask = ref z in
+
+    let rec to_pretty_bits' acc =
+      let current_bit_known = (!known_bitmask &: Ints_t.one) = Ints_t.one in
+      let current_bit_impossible = (!invalid_bitmask &: Ints_t.one) = Ints_t.one in
+
+      let bit_value = !o_mask &: Ints_t.one in
+      let bit =
+        if current_bit_impossible then "⊥"
+        else if not current_bit_known then "⊤"
+        else Ints_t.to_string bit_value 
+      in
+
+      if (!o_mask = Ints_t.of_int (-1) || !o_mask = Ints_t.zero ) && (!z_mask = Ints_t.of_int (-1) || !z_mask = Ints_t.zero) then
+        let prefix = bit ^ "..." ^ bit in
+        prefix ^ acc
+      else
+        (known_bitmask := !known_bitmask >>: 1;
+         invalid_bitmask := !invalid_bitmask >>: 1;
+         o_mask := !o_mask >>: 1;
+         z_mask := !z_mask >>: 1;
+         to_pretty_bits' (bit ^ acc))
+    in
+    "0b" ^ to_pretty_bits' ""
+
+  let show t = 
+    if t = bot () then "bot" else
+    if t = top () then "top" else
+      let (z,o) = t in
+      Format.sprintf "{%s, (zs:%s, os:%s)}" (to_pretty_bits t) (Ints_t.to_string z) (Ints_t.to_string o) 
+
+  include Std (struct type nonrec t = t let name = name let top_of = top_of let bot_of = bot_of let show = show let equal = equal end)
+
+  let range ik bf = (BArith.min ik bf, BArith.max ik bf)
+
+  let maximal (z,o) =     
+    if (z < Ints_t.zero) <> (o < Ints_t.zero) then Some o
+    else None
+
+  let minimal (z,o) =     
+    if (z < Ints_t.zero) <> (o < Ints_t.zero) then Some (!:z)
```

Comment:
Using the polymorphic compare `<` here probably works, but the compare on `Ints_t` is probably still the better choice?